File: mobile_client/lib/services/model_selection.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import 'on_device_ai_types.dart';

/// Per-surface choice of the SERVER model/provider. Chat and Voice each keep
/// their own selection so you can run (say) a fast model for voice and a
/// stronger one for chat. Persisted in secure storage, mirroring
/// [Credentials]'s pattern (Keychain on iOS, EncryptedSharedPreferences on
/// Android).
///
/// A null value for either field means "no explicit override" — the caller
/// should then fall back to the server's active/default model.
///
/// Keys:
///   - sel_chat_model     / sel_chat_provider
///   - sel_voice_model    / sel_voice_provider
class ModelSelection {
  ModelSelection._();
  static final ModelSelection instance = ModelSelection._();

  // ignore: prefer_const_declarations
  static final FlutterSecureStorage _store = const FlutterSecureStorage(
    aOptions: AndroidOptions(encryptedSharedPreferences: true),
    iOptions: IOSOptions(accessibility: KeychainAccessibility.first_unlock),
  );

  String? chatModel;
  String? chatProvider;
  String? voiceModel;
  String? voiceProvider;

  static String _modelKey(VoiceSurface s) =>
      s == VoiceSurface.voice ? 'sel_voice_model' : 'sel_chat_model';
  static String _providerKey(VoiceSurface s) =>
      s == VoiceSurface.voice ? 'sel_voice_provider' : 'sel_chat_provider';

  /// Load the persisted selections. Safe to call once at startup; the getters
  /// return null before [load] completes (i.e. read-before-load is fine).
  Future<void> load() async {
    chatModel = await _store.read(key: 'sel_chat_model');
    chatProvider = await _store.read(key: 'sel_chat_provider');
    voiceModel = await _store.read(key: 'sel_voice_model');
    voiceProvider = await _store.read(key: 'sel_voice_provider');
  }

  /// The selected model id for [surface], or null when none is set.
  String? modelFor(VoiceSurface surface) =>
      surface == VoiceSurface.voice ? voiceModel : chatModel;

  /// The selected provider id for [surface], or null when none is set.
  String? providerFor(VoiceSurface surface) =>
      surface == VoiceSurface.voice ? voiceProvider : chatProvider;

  /// Persist a new selection for [surface]. A null argument clears that field.
  ///
  /// Pass both [model] and [provider] together when picking a concrete model —
  /// they always travel as a pair from the picker.
  Future<void> setFor(
    VoiceSurface surface, {
    String? model,
    String? provider,
  }) async {
    if (surface == VoiceSurface.voice) {
      voiceModel = model;
      voiceProvider = provider;
    } else {
      chatModel = model;
      chatProvider = provider;
    }
    await _put(_modelKey(surface), model);
    await _put(_providerKey(surface), provider);
  }

  /// Clear every persisted selection (e.g. on un-pair).
  Future<void> clear() async {
    chatModel = null;
    chatProvider = null;
    voiceModel = null;
    voiceProvider = null;
    await _store.delete(key: 'sel_chat_model');
    await _store.delete(key: 'sel_chat_provider');
    await _store.delete(key: 'sel_voice_model');
    await _store.delete(key: 'sel_voice_provider');
  }

  static Future<void> _put(String key, String? value) async {
    if (value == null || value.isEmpty) {
      await _store.delete(key: key);
    } else {
      await _store.write(key: key, value: value);
    }
  }
}

File: ios_app/JarvisCopilot/Copilot/Chat/ModelSelection.swift
import Foundation

/// Per-surface choice of the SERVER model/provider. Chat and voice each keep their
/// own selection, so you can run a fast model for voice and a stronger one for
/// chat. Ported from `services/model_selection.dart`.
///
/// `VoiceSurface` (chat | voice) is the shared enum from `Copilot/Skills` — the
/// same name Flutter uses, and what the voice surface already keys off.
///
/// A nil value for either field means "no explicit override" — the caller then
/// falls back to the server's active/default model.
///
/// Keys (unchanged from the Flutter app, so a migrated install keeps its choice):
///   - `sel_chat_model`  / `sel_chat_provider`
///   - `sel_voice_model` / `sel_voice_provider`
struct ModelSelection: Sendable {
    /// Preferences, not credentials: the model id is not a secret, and reading it
    /// synchronously means the composer never renders a stale "Default".
    let store: KeyValueStore

    static let shared = ModelSelection(store: UserDefaults.standard)

    init(store: KeyValueStore = UserDefaults.standard) { self.store = store }

    func model(for surface: VoiceSurface) -> String? { read(Self.modelKey(surface)) }
    func provider(for surface: VoiceSurface) -> String? { read(Self.providerKey(surface)) }

    /// Persist a new selection for `surface`. A nil (or empty) argument clears that
    /// field. Model and provider always travel as a pair from the picker.
    func set(_ surface: VoiceSurface, model: String?, provider: String?) {
        write(Self.modelKey(surface), model)
        write(Self.providerKey(surface), provider)
    }

    /// Clear every persisted selection (e.g. on un-pair).
    func clear() {
        set(.chat, model: nil, provider: nil)
        set(.voice, model: nil, provider: nil)
    }

    private func read(_ key: String) -> String? {
        guard let value = store.string(key), !value.isEmpty else { return nil }
        return value
    }

    private func write(_ key: String, _ value: String?) {
        store.set((value?.isEmpty == false) ? value : nil, forKey: key)
    }

    private static func modelKey(_ surface: VoiceSurface) -> String { "sel_\(surface.rawValue)_model" }
    private static func providerKey(_ surface: VoiceSurface) -> String { "sel_\(surface.rawValue)_provider" }
}
